amcat4apiclient/auth.py

```diff
@@ -0,0 +1,180 @@
+import requests
+import os
+from appdirs import user_cache_dir
+from base64 import urlsafe_b64encode, b64encode
+from cryptography.fernet import Fernet
+from cryptography.fernet import Fernet
+from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
+from datetime import datetime, timedelta
+from hashlib import sha256
+from json import dumps, loads
+from random import getrandbits
+from re import search
+from requests_oauthlib import OAuth2Session
+from socket import socket, AF_INET, SOCK_STREAM
+from webbrowser import open as browse
+
+
+def get_middlecat_token(host, callback_port=65432, refresh = "static"):
+    middlecat = requests.get(f"{host}/middlecat").json()["middlecat_url"]
+    auth_url = f"{middlecat}/authorize" 
+    token_url = f"{middlecat}/api/token"
+    pkce = pkce_challange()
+
+    auth_params = {
+        "resource": host, 
+        "refresh": refresh, 
+        "session_type": "api_key",
+        "code_challenge_method": pkce["method"],
+        "code_challenge": pkce["challenge"]
+    }
+
+    oauth = OAuth2Session(client_id="amcat4apiclient", redirect_uri=f"http://localhost:{callback_port}/")
+
+    authorization_url, state = oauth.authorization_url(auth_url, **auth_params)
+
+    # open a socket for callbacks
+    s = socket(AF_INET, SOCK_STREAM)
+    s.bind(("127.0.0.1", callback_port))
+    s.listen()
+
+    # open browser for user
+    browse(authorization_url)
+
+    # wait for get request with code
+    print("Waiting for authorization in browser...")
```

Comment:
I don't like print statements in library code, would be better to replace by logging

Reply:
(although I guess this is not really meant as a library function as it requires interactivity anyways?)

Reply:
This mimics the process in R where the browser is opened with the correct URL for middlecat authentication if no token is found. The message here is for the user to understand that input is required before authentication can continue. Logging would not give the user this hint, which could lead them to believe Python is still processing something and eventually abort the process. For library vs object-oriented see below.